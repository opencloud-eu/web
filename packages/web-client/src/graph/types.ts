import type { AxiosInstance, ResponseType } from 'axios'
import type { Configuration } from './generated'

export interface GraphFactoryOptions {
  axiosClient: AxiosInstance
  config: Configuration
}

export interface GraphRequestOptions {
  headers?: Record<string, string>
  params?: Record<string, string>
  signal?: AbortSignal
  responseType?: ResponseType
}
