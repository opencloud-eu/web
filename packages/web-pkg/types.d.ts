import { Ability } from '@opencloud-eu/web-client'
import {
  ArchiverService,
  ClientService,
  LoadingService,
  PreviewService,
  PasswordPolicyService
} from './src/services'
import { Route, Router } from 'vue-router'

export * from './src'

declare module 'vue' {
  interface ComponentCustomProperties {
    $ability: Ability
    $archiverService: ArchiverService
    $clientService: ClientService
    $loadingService: LoadingService
    $previewService: PreviewService
    $passwordPolicyService: PasswordPolicyService
    $router: Router
    $route: Route
  }
}
