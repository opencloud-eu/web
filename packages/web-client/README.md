# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
