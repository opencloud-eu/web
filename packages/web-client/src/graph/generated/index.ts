/* tslint:disable */
/* eslint-disable */
/**
 * Libre Graph API
 * Libre Graph is a free API for cloud collaboration inspired by the MS Graph API.
 *
 * The version of the OpenAPI document: v1.0.7
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export * from "./api";
export * from "./configuration";

