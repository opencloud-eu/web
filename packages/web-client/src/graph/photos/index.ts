export * from './photo'
export * from './types'
