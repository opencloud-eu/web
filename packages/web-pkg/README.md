# web-pkg

`web-pkg` is a package that provides utilities, most importantly a variety of components and composables, that can be useful when developing apps and extensions for OpenCloud Web.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-pkg --save-dev

$ pnpm add -D @opencloud-eu/web-pkg

$ yarn add @opencloud-eu/web-pkg --dev
```

It's recommended to install this package as a dev dependency because it's only really needed for providing autocompletion in your IDE and unit tests. In a runtime context, the OpenCloud Web runtime provides the actual implementation.
