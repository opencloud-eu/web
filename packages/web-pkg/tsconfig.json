{
  "extends": "@opencloud-eu/tsconfig",

  // necessary for building web-pkg standalone, otherwise we get errors with inferred types
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@vue/shared": ["node_modules/@vue/shared/"],
      "@codemirror/autocomplete": ["./node_modules/@codemirror/autocomplete"],
      "vue/jsx-runtime": ["./node_modules/vue/jsx-runtime"],
      "@popperjs/core": ["./../design-system/node_modules/@popperjs/core"]
    }
  },

  // include component types of the design-system
  "include": ["../design-system/types.d.ts"]
}
