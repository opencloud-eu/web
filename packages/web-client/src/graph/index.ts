import { AxiosInstance } from 'axios'
import { Configuration } from './generated'
import { type GraphUsers, UsersFactory } from './users'
import { type GraphGroups, GroupsFactory } from './groups'
import { ApplicationsFactory, GraphApplications } from './applications'
import { DrivesFactory, GraphDrives } from './drives'
import { DriveItemsFactory, GraphDriveItems } from './driveItems'
import { TagsFactory, GraphTags } from './tags'
import { ActivitiesFactory, GraphActivities } from './activities'
import { PermissionsFactory, GraphPermissions } from './permissions'
import { PhotoFactory, GraphPhotos } from './photos'

export interface Graph {
  activities: GraphActivities
  applications: GraphApplications
  tags: GraphTags
  drives: GraphDrives
  driveItems: GraphDriveItems
  users: GraphUsers
  groups: GraphGroups
  permissions: GraphPermissions
  photos: GraphPhotos
}

export const graph = (baseURI: string, axiosClient: AxiosInstance): Graph => {
  const url = new URL(baseURI)
  url.pathname = [...url.pathname.split('/'), 'graph'].filter(Boolean).join('/')
  const config = new Configuration({
    basePath: url.href
  })

  return <Graph>{
    activities: ActivitiesFactory({ axiosClient, config }),
    applications: ApplicationsFactory({ axiosClient, config }),
    tags: TagsFactory({ axiosClient, config }),
    drives: DrivesFactory({ axiosClient, config }),
    driveItems: DriveItemsFactory({ axiosClient, config }),
    users: UsersFactory({ axiosClient, config }),
    groups: GroupsFactory({ axiosClient, config }),
    permissions: PermissionsFactory({ axiosClient, config }),
    photos: PhotoFactory({ axiosClient, config })
  }
}
