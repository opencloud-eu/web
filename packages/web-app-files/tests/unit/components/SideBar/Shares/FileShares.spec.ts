import FileShares from '../../../../../src/components/SideBar/Shares/FileShares.vue'
import { mock } from 'vitest-mock-extended'
import { Resource } from '@opencloud-eu/web-client'
import { SpaceResource } from '@opencloud-eu/web-client'
import { v4 as uuidV4 } from 'uuid'
import { CollaboratorShare, ShareRole, ShareTypes } from '@opencloud-eu/web-client'
import {
  defaultPlugins,
  mount,
  shallowMount,
  defaultComponentMocks,
  defaultStubs
} from '@opencloud-eu/web-test-helpers'
import CollaboratorListItem from '../../../../../src/components/SideBar/Shares/Collaborators/ListItem.vue'
import { AncestorMetaData, CapabilityStore, useCanShare, useModals } from '@opencloud-eu/web-pkg'
import { User } from '@opencloud-eu/web-client/graph/generated'

vi.mock('@opencloud-eu/web-pkg', async (importOriginal) => ({
  ...(await importOriginal<any>()),
  useCanShare: vi.fn()
}))

const getCollaborator = (): CollaboratorShare => ({
  id: uuidV4(),
  sharedWith: {
    id: 'einstein',
    displayName: 'Albert Einstein'
  },
  sharedBy: {
    id: 'admin',
    displayName: 'Admin'
  },
  permissions: [],
  role: mock<ShareRole>(),
  resourceId: uuidV4(),
  indirect: false,
  shareType: ShareTypes.user.value,
  createdDateTime: '2024-01-01'
})

describe('FileShares', () => {
  describe('invite collaborator form', () => {
    it('renders the form when the resource can be shared', () => {
      const resource = mock<Resource>()
      const { wrapper } = getWrapper({ resource, canShare: true })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeTruthy()
    })
    it('does not render the form when the resource can not be shared', () => {
      const resource = mock<Resource>()
      const { wrapper } = getWrapper({ resource, canShare: false })
      expect(wrapper.find('invite-collaborator-form-stub').exists()).toBeFalsy()
    })
  })

  describe('collaborators list', () => {
    let collaborators: CollaboratorShare[]
    beforeEach(() => {
      collaborators = [getCollaborator(), getCollaborator(), getCollaborator(), getCollaborator()]
    })

    it('renders sharedWithLabel and sharee list', async () => {
      const { wrapper } = getWrapper({ collaborators })
      await wrapper.find('.toggle-shares-list-btn').trigger('click')
      expect(wrapper.find('#files-collaborators-list').exists()).toBeTruthy()
      expect(wrapper.findAll('#files-collaborators-list li').length).toBe(collaborators.length)
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('reacts on delete events', async () => {
      const { wrapper } = getWrapper({ collaborators })
      const { dispatchModal } = useModals()
      wrapper
        .findComponent<typeof CollaboratorListItem>('collaborator-list-item-stub')
        .vm.$emit('onDelete')
      await wrapper.vm.$nextTick()
      expect(dispatchModal).toHaveBeenCalledTimes(1)
    })
    it('correctly passes the shared parent route to the collaborator list item for indirect shares', () => {
      const indirectCollaborator = { ...getCollaborator(), indirect: true }
      const ancestorMetaData = {
        '/somePath': { id: indirectCollaborator.resourceId }
      } as unknown as AncestorMetaData
      const { wrapper } = getWrapper({ collaborators: [indirectCollaborator], ancestorMetaData })
      const listItemStub = wrapper.findComponent<typeof CollaboratorListItem>(
        'collaborator-list-item-stub'
      )
      expect(listItemStub.props('sharedParentRoute')).toBeTruthy()
      expect(listItemStub.props('modifiable')).toBeFalsy()
    })
    it('toggles the share list', async () => {
      const { wrapper } = getWrapper({ mountType: mount, collaborators })
      expect(wrapper.vm.sharesListCollapsed).toBe(true)
      await wrapper.find('.toggle-shares-list-btn').trigger('click')
      expect(wrapper.vm.sharesListCollapsed).toBe(false)
    })
    it('share should be modifiable if its personal space share', () => {
      const space = mock<SpaceResource>({ driveType: 'personal' })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(true)
    })
    it('share should not be modifiable if collaborator is indirect', () => {
      const space = mock<SpaceResource>({ driveType: 'personal' })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      collaborators[0]['indirect'] = true
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(false)
    })
    it('share should not be modifiable if its a project space user cannot share', () => {
      const space = mock<SpaceResource>({ driveType: 'project', canShare: () => false })
      collaborators[0]['indirect'] = true
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(false)
    })
    it('share should be modifiable if its a project space and user can share', () => {
      const space = mock<SpaceResource>({ driveType: 'project', canShare: () => true })
      const { wrapper } = getWrapper({ space, mountType: shallowMount, collaborators })
      expect(wrapper.vm.isShareModifiable(collaborators[0])).toBe(true)
    })
  })

  describe('current space', () => {
    it('loads space members if a space is given and the current user is member', () => {
      const user = { id: '1' } as User
      const space = mock<SpaceResource>({ driveType: 'project' })
      const spaceMembers = [
        { sharedWith: { id: user.id, displayName: '' }, resourceId: space.id, permissions: [] },
        { sharedWith: { id: '2', displayName: '' }, resourceId: space.id, permissions: [] }
      ] as CollaboratorShare[]
      const collaborator = getCollaborator()
      collaborator.sharedWith = {
        ...collaborator.sharedWith,
        id: user.id
      }
      const { wrapper } = getWrapper({ space, collaborators: [collaborator], user, spaceMembers })
      expect(wrapper.find('#files-collaborators-list').exists()).toBeTruthy()
      expect(wrapper.findAll('#files-collaborators-list li').length).toBe(1)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('"deleteShareConfirmation" method', () => {
    it('dispatches a modal', () => {
      const { wrapper } = getWrapper()
      const { dispatchModal } = useModals()
      wrapper.vm.deleteShareConfirmation(mock<CollaboratorShare>())
      expect(dispatchModal).toHaveBeenCalled()
    })
  })
})

function getWrapper({
  mountType = shallowMount,
  resource = mock<Resource>({ isReceivedShare: () => false, canShare: () => true }),
  space = mock<SpaceResource>(),
  collaborators = [],
  spaceMembers = [],
  user = undefined,
  ancestorMetaData = {},
  canShare = true
}: {
  mountType?: typeof mount
  resource?: Resource
  space?: SpaceResource
  collaborators?: CollaboratorShare[]
  spaceMembers?: CollaboratorShare[]
  user?: User
  ancestorMetaData?: AncestorMetaData
  canShare?: boolean
} = {}) {
  vi.mocked(useCanShare).mockReturnValue({ canShare: () => canShare })

  const capabilities = {
    files_sharing: { deny_access: false }
  } satisfies Partial<CapabilityStore['capabilities']>

  if (spaceMembers.length) {
    collaborators = [...collaborators, ...spaceMembers]
  }

  return {
    wrapper: mountType(FileShares, {
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              stubActions: false,
              userState: { user },
              capabilityState: { capabilities },
              configState: {
                options: { contextHelpers: true }
              },
              sharesState: { collaboratorShares: collaborators },
              resourcesStore: { ancestorMetaData }
            }
          })
        ],
        mocks: defaultComponentMocks(),
        provide: { resource, space },
        stubs: {
          ...defaultStubs,
          OcButton: false,
          'invite-collaborator-form': true,
          'collaborator-list-item': true
        }
      }
    })
  }
}
