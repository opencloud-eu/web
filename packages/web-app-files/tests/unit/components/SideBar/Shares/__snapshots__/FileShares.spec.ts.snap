// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileShares > collaborators list > renders sharedWithLabel and sharee list 1`] = `
"<div id="oc-files-sharing-sidebar" class="relative">
  <div class="flex justify-between items-center">
    <div class="flex">
      <h3 class="font-semibold text-base m-0" data-msgid="Share with people" data-current-language="en">Share with people</h3>
      <oc-contextual-helper-stub title="Share with people" endtext="" list="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" readmorelink="https://docs.opencloud.eu/docs/user/sharing" text="Use the input field to search for users and groups. Select them to share the item." class="pl-1"></oc-contextual-helper-stub>
    </div>
    <copy-private-link-stub resource="undefined"></copy-private-link-stub>
  </div>
  <invite-collaborator-form-stub savebuttonlabel="Share" invitelabel="" class="my-2"></invite-collaborator-form-stub>
  <div id="files-collaborators-headline" class="flex items-center justify-between h-10">
    <h4 class="font-semibold my-0">Shared with</h4>
  </div>
  <portal-target name="app.files.sidebar.sharing.shared-with.top" slot-props="[object Object]" multiple="true"></portal-target>
  <ul id="files-collaborators-list" class="oc-list oc-list-divider m-0" aria-label="Share receivers">
    <li>
      <collaborator-list-item-stub share="[object Object]" issharedenied="false" modifiable="true" removable="true" resourcename="[Function]" deniable="false" islocked="[Function]" isspaceshare="false"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" issharedenied="false" modifiable="true" removable="true" resourcename="[Function]" deniable="false" islocked="[Function]" isspaceshare="false"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" issharedenied="false" modifiable="true" removable="true" resourcename="[Function]" deniable="false" islocked="[Function]" isspaceshare="false"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" issharedenied="false" modifiable="true" removable="true" resourcename="[Function]" deniable="false" islocked="[Function]" isspaceshare="false"></collaborator-list-item-stub>
    </li>
    <portal-target name="app.files.sidebar.sharing.shared-with.bottom" slot-props="[object Object]" multiple="true"></portal-target>
  </ul>
  <div class="flex justify-center"><button type="button" class="oc-button-secondary oc-button-raw oc-button-secondary-raw gap-2 justify-center text-base min-h-4 oc-button inline-flex toggle-shares-list-btn">
      <!--v-if-->
      <!-- @slot Content of the button --> Show less
    </button></div>
  <!--v-if-->
</div>"
`;

exports[`FileShares > current space > loads space members if a space is given and the current user is member 1`] = `
"<div id="oc-files-sharing-sidebar" class="relative">
  <div class="flex justify-between items-center">
    <div class="flex">
      <h3 class="font-semibold text-base m-0" data-msgid="Share with people" data-current-language="en">Share with people</h3>
      <oc-contextual-helper-stub title="Share with people" endtext="" list="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" readmorelink="https://docs.opencloud.eu/docs/user/sharing" text="Use the input field to search for users and groups. Select them to share the item." class="pl-1"></oc-contextual-helper-stub>
    </div>
    <copy-private-link-stub resource="undefined"></copy-private-link-stub>
  </div>
  <invite-collaborator-form-stub savebuttonlabel="Share" invitelabel="" class="my-2"></invite-collaborator-form-stub>
  <div id="files-collaborators-headline" class="flex items-center justify-between h-10">
    <h4 class="font-semibold my-0">Shared with</h4>
  </div>
  <portal-target name="app.files.sidebar.sharing.shared-with.top" slot-props="[object Object]" multiple="true"></portal-target>
  <ul id="files-collaborators-list" class="oc-list oc-list-divider mb-6" aria-label="Share receivers">
    <li>
      <collaborator-list-item-stub share="[object Object]" issharedenied="false" modifiable="false" removable="false" resourcename="[Function]" deniable="false" islocked="[Function]" isspaceshare="false"></collaborator-list-item-stub>
    </li>
    <portal-target name="app.files.sidebar.sharing.shared-with.bottom" slot-props="[object Object]" multiple="true"></portal-target>
  </ul>
  <!--v-if-->
  <div class="flex items-center justify-between">
    <h4 class="font-semibold my-2">Space members</h4>
  </div>
  <ul id="space-collaborators-list" class="oc-list oc-list-divider overflow-hidden m-0" aria-label="Space members">
    <li>
      <collaborator-list-item-stub share="[object Object]" issharedenied="false" modifiable="false" removable="false" resourcename="[Function]" deniable="false" islocked="false" isspaceshare="true"></collaborator-list-item-stub>
    </li>
    <li>
      <collaborator-list-item-stub share="[object Object]" issharedenied="false" modifiable="false" removable="false" resourcename="[Function]" deniable="false" islocked="false" isspaceshare="true"></collaborator-list-item-stub>
    </li>
  </ul>
  <!--v-if-->
</div>"
`;
